tests/ipsecondary_test.py

```diff
@@ -0,0 +1,537 @@
+    logger.debug("Update result: {}".format("success" if test_result else "failed"))
+    return test_result
+
+
+@with_uptane_backend()
+@with_secondary(start=False)
+@with_aktualizr(start=False, output_logs=False)
+def test_secondary_update(uptane_server, secondary, aktualizr):
+    '''Test Secondary update if a boot order of Secondary and Primary is undefined'''
+
+    test_result = True
+    number_of_updates = 1
+    ii = 0
+    while ii < number_of_updates and test_result:
+        # add a new image to the repo in order to update the secondary with it
+        secondary_image_filename = "secondary_image_filename_{}.img".format(ii)
+        secondary_image_hash = uptane_server.repo.add_image(id=secondary.id, image_filename=secondary_image_filename)
+
+        logger.debug("Trying to update ECU {} with the image {}".
+                    format(secondary.id, (secondary_image_hash, secondary_image_filename)))
+
+        # start Secondary and Aktualizr processes, aktualizr is started in 'once' mode
+        with secondary, aktualizr:
+            aktualizr.wait_for_completion()
+
+        test_result = (secondary_image_hash, secondary_image_filename) == aktualizr.get_current_image_info(secondary.id)
+        logger.debug("Update result: {}".format("success" if test_result else "failed"))
+        ii += 1
+
+    return test_result
+
+
+@with_uptane_backend()
+@with_secondary(start=False)
+@with_aktualizr(start=False, output_logs=False, wait_timeout=0.1)
+def test_primary_timeout_during_first_run(uptane_server, secondary, aktualizr):
+    '''Test Aktualizr's timeout of waiting for Secondaries during initial boot'''
+
+    secondary_image_filename = "secondary_image_filename_001.img"
+    secondary_image_hash = uptane_server.repo.add_image(id=secondary.id, image_filename=secondary_image_filename)
+
+    logger.debug("Checking Aktualizr behaviour if it timeouts while waiting for a connection from the secondary")
+
+    # just start the aktualizr and expect that it timeouts on waiting for a connection from the secondary
+    # so the secondary is not registered at the device and backend
+    with aktualizr:
+        aktualizr.wait_for_completion()
+
+    return not aktualizr.is_ecu_registered(secondary.id)
+
+
+@with_uptane_backend()
+@with_secondary(start=False)
+@with_aktualizr(start=False, output_logs=False)
+def test_primary_timeout_after_device_is_registered(uptane_server, secondary, aktualizr):
+    '''Test Aktualizr's timeout of waiting for Secondaries after the device/aktualizr was registered at the backend'''
+
+    # run aktualizr and secondary and wait until the device/aktualizr is registered
+    with aktualizr, secondary:
+        aktualizr.wait_for_completion()
+
+    # the secondary must be registered
+    if not aktualizr.is_ecu_registered(secondary.id):
+        return False
+
+    # run just aktualizr, the previously registered secondary is off
+    # and check if the primary ECU is updatable if the secondary is not connected
```

Comment:
How do you control/confirm that the secondary is not running here?

Reply:
the context manager of the 'with' statement at line 486 guarantees that IPSecondary::__exit__ is called, but I'll add additional check to make sure that the secondary's process is not running.

Reply:
Just updated